src/Orleans.Runtime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceActivationMaintainer.cs

```diff
@@ -95,10 +101,21 @@ public void TrackDoubtfulGrains(Dictionary<GrainId, IGrainInfo> newstuff)
             {
                 try
                 {
-                    await Task.Delay(period);
-                    if (!router.Running) break;
+                    // wait until it is time, or someone prodded us to continue immediately
+                    try
+                    {
+                        await Task.Delay(period, cts.Token);
 
-                    logger.Debug("GSIP:M running periodic check (having waited {0})", period);
+                        logger.Debug("GSIP:M running periodic check (having waited {0})", period);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        logger.Debug("GSIP:M running check");
+
+                        cts = new CancellationTokenSource(); // reset for next time
```

Comment:
There's also a protected `Cts` member which is in scope here. The two might be misleading. `Cts` will be cancelled when the agent is supposed to terminate.

Reply:
that's definitely confusing. I will choose some better name.